### ISSUE TITLE
linux cubeb pipewire jack: any game stalls when trying to open audio playback device

### ISSUE THREAD

**Issue by mazes-80** [cemu-project/Cemu#601]
Running with one pipewire global instance cubeb detects one 'Jack Playback' audio output available but:

[14:40:12.328] Cubeb: available
[14:40:12.328] ------- Init Audio Input backend -------
[14:40:12.328] Cubeb: available
[14:40:12.328] ------- Run title -------
[14:40:12.879] can't initialize tv audio: can't initialize cubeb device

Then the emulator stalls.

Running under pure alsa I got no audio output device listed as available.

does the emulator require pulseaudio to emit sound ?

Everything else seems to work fine and I got way better rendering than using 'wine Cemu.exe'

**Comment by Squall-Leonhart** [comment]
Refer to comments in https://github.com/cemu-project/Cemu/issues/428

**Comment by mazes-80** [comment]
> Refer to comments in #428
Are you referring: 
@jn64 Thanks for the feedback. I had deleted my post, before your response, since I figured that I had to manually select the audio device under TV. It's working great now. +1

I forgot to mention I already selected the device, but log are clearly different from other thread: "can't initialize tv audio: can't initialize cubeb device" -> device selected: not able to open it -> hang.

If I select no audio output, I can play perfect, without sound but nothing hangs, and log.txt becomes similar to other thread.

**Comment by mazes-80** [comment]
It seems I can't use pipewire 'Jack Playback' output, that pure alsa does not return any output available.
Comment from other thread made me think pulseaudio is maybe the only way to get cemu sound working now.
I never used pulseaudio and may never use it despite some very appealing features.
I may look a bit deeper at other software way of handling various cubeb outputs if someone can confirm for now pulseaudio is mandatory under linux for cemu.
From what I could see in current code, nothing seems especially related to pulseaudio in cemu looping through cubeb outputs.

**Comment by mazes-80** [comment]
My build for cemu is not standard as I bundled libfmt. (should not matter)
in gentoo I can't install libfmt>9 without dropping some other package now
I still didn't understand where to install resources when doing system shared install with PORTABLE=OFF, shouldn't matter too much also, just fonts and translation are missing (I will for sure look deeper at this)

**Comment by jn64** [comment]
What do you mean by 'Jack Playback' output? Are you using pipewire-jack?

pulseaudio isn't strictly required for Cubeb, but the [ALSA/JACK support may not be up to par](https://github.com/mozilla/cubeb/wiki/Backend-Support). I don't know the actual current status in practice. I'm also not familiar with Gentoo, so you might have to check how your Cubeb is compiled. 

But since you use pipewire, is there a reason you don't want to use pipewire-pulse? Very few applications use pipewire directly, while many use pulseaudio API, so having pipewire-pulse isn't just relevant for Cemu/Cubeb.

See also:

- [Does PipeWire run on top of a PulseAudio or JACK server? | FAQ · Wiki · PipeWire / pipewire · GitLab](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#does-pipewire-run-on-top-of-a-pulseaudio-or-jack-server)
- [Should I uninstall everything pulseaudio? | FAQ · Wiki · PipeWire / pipewire · GitLab](https://gitlab.freedesktop.org/pipewire/pipewire/-/wikis/FAQ#should-i-uninstall-everything-pulseaudio)
  > Some applications (firefox) can also be compiled with different backends but this
is not recommended because:
  > The PulseAudio backend is usually the best (only?) maintained one.

---

> where to install resources when doing system shared install with PORTABLE=OFF

See https://github.com/cemu-project/Cemu/pull/316#issue-1392163125 and https://github.com/cemu-project/Cemu/pull/316#issuecomment-1263570866.

Basically, if your Cemu binary is `/usr/bin/Cemu`, then:

- `bin/resources` should be installed to `/usr/share/Cemu/resources`
- `bin/gameProfiles` should be installed to `/usr/share/Cemu/gameProfiles`

**Comment by mazes-80** [comment]
'Jack Playback': the name listed by cubeb relayed to cemu gui. provided by pipewire jack as suggested in title.

I never installed pulseaudio because bad jack-audio-connection-kit cohabitation. 
Now with pipewire, I got no real blockers for pulseaudio. It may lead me to have working sound in cemu and a lot of other things to fix for all my other setup or just work well. Like when moving to pipewire configuration can be time consuming (issues with midi and mpd)

About /usr/share/Cemu, i was so close: /usr/share/cemu. Thank you

**Comment by jn64** [comment]
Sorry, I missed the title.

I can reproduce the issue on Fedora 37 with pipewire-jack (and no pulse server running). Same log, emulator stalls. So at least it's not unique to your Cemu/Cubeb build.

I understand about old pulseaudio/JACK conflicts, but like you said, pipewire solves it. I think as far as Cemu is concerned, the only practical advice is to use pulseaudio/pipewire-pulse. It's unlikely anyone here can do anything about this issue, whether the problem lies with Cubeb's JACK backend, or possibly with pipewire-jack.

**Comment by mazes-80** [comment]
Nice to read your last comment.
I may try to take a look at how other emulators handles cubeb.
If it end up with something, even just working for me, or not so good quality it'll be PR'd
Really thanks for the effort you put in tracking this issue.

**Comment by mazes-80** [comment]
Also I'm contributor to a gentoo pool of new packages: guru overlay.
Package is now ready locally but I need to wait wxGTK>=3.2 to be supported in gentoo before dropping it to guru.
I hope this package to be available under gentoo soon.

**Comment by goeiecool9999** [comment]
Cemu is programmed to crash when it can't initialize a selected audio device. The problem is that it does so with ``exit``. Which behaves differently on Linux compared to Windows. On windows it'll actually shut down the whole process. On linux it only shuts down the calling thread. In this case it's the thread responsible for starting up the game. That explains the hang. When porting from Windows to Linux ``exit`` was changed to ``_Exit`` (which behaves the same as Windows) in some places but not others.
